devito/dle/blocking_utils.py

```diff
@@ -1,18 +1,114 @@
+                # won't be computationally expensive. This will help with code
+                # size/redability, JIT time, and auto-tuning time
+                if not (tree.root.is_Sequential or iet.is_Callable):
+                    # E.g., not inside a time-stepping Iteration
+                    continue
+                if any(i.dim.is_Sub and i.dim.local for i in tree):
+                    # At least an outer Iteration is over a local SubDimension,
+                    # which suggests the computational cost of this Iteration
+                    # nest will be negligible w.r.t. the "core" Iteration nest
+                    # (making use of non-local (Sub)Dimensions only)
+                    continue
+            if not IsPerfectIteration().visit(root):
+                # Don't know how to block non-perfect nests
+                continue
+
+            # Apply loop blocking to `tree`
+            interb = []
+            intrab = []
+            for i in iterations:
+                d = BlockDimension(i.dim, name="%s%d_blk" % (i.dim.name, self.nblocked))
+                block_dims.append(d)
+                # Build Iteration over blocks
+                properties = (PARALLEL,) + ((AFFINE,) if i.is_Affine else ())
+                interb.append(Iteration([], d, d.symbolic_max, properties=properties))
+                # Build Iteration within a block
+                intrab.append(i._rebuild([], limits=(d, d+d.step-1, 1), offsets=(0, 0)))
+
+            # Construct the blocked tree
+            blocked = compose_nodes(interb + intrab + [iterations[-1].nodes])
+            blocked = unfold_blocked_tree(blocked)
+
+            # Promote to a separate Callable
+            dynamic_parameters = flatten((bi.dim, bi.dim.symbolic_size) for bi in interb)
+            efunc = make_efunc("bf%d" % self.nblocked, blocked, dynamic_parameters)
+            efuncs.append(efunc)
+
+            # Compute the iteration ranges
+            ranges = []
+            for i, bi in zip(iterations, interb):
+                maxb = i.symbolic_max - (i.symbolic_size % bi.dim.step)
+                ranges.append(((i.symbolic_min, maxb, bi.dim.step),
+                               (maxb + 1, i.symbolic_max, i.symbolic_max - maxb)))
+
+            # Build Calls to the `efunc`
+            body = []
+            for p in product(*ranges):
+                dynamic_args_mapper = {}
+                for bi, (m, M, b) in zip(interb, p):
+                    dynamic_args_mapper[bi.dim] = (m, M)
+                    dynamic_args_mapper[bi.dim.step] = (b,)
+                call = efunc.make_call(dynamic_args_mapper)
+                body.append(List(body=call))
+
+            mapper[root] = List(body=body)
+
+            # Next blockable nest, use different (unique) variable/function names
+            self.nblocked += 1
+
+        iet = Transformer(mapper).visit(iet)
+
+        # Force-unfold if some folded Iterations haven't been blocked in the end
```

Comment:
Could you elaborate more on this docstring?

Reply:
I tried, but I can't find an easy way. Will do another day